# ![QIAGEN logo](https://www.qiagen.com/sfc/images/qiagen-logo.png) Welcome QIAGEN team!

Following up on my application for QDI's [Senior Bioinformatics Scientist](https://www.qiagen.com/us/about-us/careers/jobs/details?jobId=17775&jobTitle=Senior%20Bioinformatics%20Scientist%2C%20QDI), Jacquey asked me to provide code demonstrating expertise with Python and R in an omics context. Jacquey also asked that the code be relavant to a pipeline I've written, or benchmarking/testing methods. I think these are _very_ sensible things to ask. Unfortunatly, nearly all of the bioinformatics code I've written recently belongs to my former employer. So instead, I've decided to write a small pipeline to address a question that has in the back of my mind since I came accross this MAQC paper relating to [single cell mRNA-seq benchmarking](#https://pubmed.ncbi.nlm.nih.gov/33349700/).

You may be familiar with computation tools used to estimate the proportion of cell types that contributed to a bulk mRNA-seq sample. These tools usually realy on data enriched for a particualr cell type or identification of cell type specific gene expression patterns from single cell mRNA-seq data. There are a number of tools available to deconvolute bulk mRNA-seq data, so I've been wondering which methods work the best and if the ratio of cells present in a mixture affects the accuracy of any deconvolution approach. Existing literature on [this topic](https://www.nature.com/articles/s41467-020-19015-1) is valuable, but relies on creating 'pseudo-bulk' mRNA-seq samples and then comparing predictions made using the single cell data that created the pseudo-bulk data.

I would like to know how deconvolution methods perform when single cell and bulk libraries are prepared independently, sladkfj;sdkla fj asdl;fksadl;fjk 



There is literature on deconvolution methods already, however, often I see  

If you would like to go straight to some Python/R code, please checkout (those links, respectivley).

If you would like to 

### Table of Contents
