{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "24c516d1-ad8f-46e1-868a-045c08382577",
   "metadata": {},
   "source": [
    "# A notebook to create convoluted bulk mRNA-seq samples for processing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "b6f29aab-b8b4-43d9-ace7-94a0b6bcdffb",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from os.path import join\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import ipynb\n",
    "import subprocess\n",
    "from glob import glob\n",
    "\n",
    "from ipynb.fs.full.python_cmd_tools import *"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "70aa243f-1b42-4048-a7d6-ab624ad88b5e",
   "metadata": {},
   "source": [
    "# load sample meta data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "47ad2a7e-e773-4a49-af47-b2fc8d9e3590",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Env variables\n",
    "envs = [line.strip().split() for line in open('./env_paths')]\n",
    "envs = {i[0]: i[1] for i in envs}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "d1d1427b-13f9-4336-97c9-2595e4365922",
   "metadata": {},
   "outputs": [],
   "source": [
    "# generated by NextFlow in fist step\n",
    "sample_sheet = join(envs['MAQC_DATA'], 'samplesheet/samplesheet.csv')\n",
    "sample_sheet = pd.read_csv(sample_sheet)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "8dddca45-8d99-4c57-b6e8-2ffb32249f81",
   "metadata": {},
   "outputs": [],
   "source": [
    "# select the .fastq files that are bulk\n",
    "bulk_samples = sample_sheet[sample_sheet.experiment_title.str.contains('bulk')]\n",
    "bulk_samples = bulk_samples.reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "bb10cc3b-abb4-46ae-a631-959950eebbd1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_title</th>\n",
       "      <th>experiment_title</th>\n",
       "      <th>run_alias</th>\n",
       "      <th>library_source</th>\n",
       "      <th>read_count</th>\n",
       "      <th>sample_description</th>\n",
       "      <th>description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HCC1395 BL</td>\n",
       "      <td>NextSeq 550 sequencing; bulk RNA-seq of HCC139...</td>\n",
       "      <td>HCC1395BL_1_S10_R1_001.fastq.gz</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>26549469</td>\n",
       "      <td>HCC1395 BL</td>\n",
       "      <td>NextSeq 550 sequencing; bulk RNA-seq of HCC139...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HCC1395 BL</td>\n",
       "      <td>NextSeq 550 sequencing; bulk RNA-seq of HCC139...</td>\n",
       "      <td>HCC1395BL_2_S11_R1_001.fastq.gz</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>28768620</td>\n",
       "      <td>HCC1395 BL</td>\n",
       "      <td>NextSeq 550 sequencing; bulk RNA-seq of HCC139...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HCC1395 BL</td>\n",
       "      <td>NextSeq 550 sequencing; bulk RNA-seq of HCC139...</td>\n",
       "      <td>HCC1395BL_3_S12_R1_001.fastq.gz</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>29246313</td>\n",
       "      <td>HCC1395 BL</td>\n",
       "      <td>NextSeq 550 sequencing; bulk RNA-seq of HCC139...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HCC1395</td>\n",
       "      <td>NextSeq 550 sequencing; bulk RNA-seq of HCC139...</td>\n",
       "      <td>HCC1395_2_S14_R1_001.fastq.gz</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>28095730</td>\n",
       "      <td>HCC1395</td>\n",
       "      <td>NextSeq 550 sequencing; bulk RNA-seq of HCC139...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HCC1395</td>\n",
       "      <td>NextSeq 550 sequencing; bulk RNA-seq of HCC139...</td>\n",
       "      <td>HCC1395_3_S15_R1_001.fastq.gz</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>27883943</td>\n",
       "      <td>HCC1395</td>\n",
       "      <td>NextSeq 550 sequencing; bulk RNA-seq of HCC139...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>HCC1395</td>\n",
       "      <td>NextSeq 550 sequencing; bulk RNA-seq of HCC139...</td>\n",
       "      <td>HCC1395_1_S13_R1_001.fastq.gz</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>30200567</td>\n",
       "      <td>HCC1395</td>\n",
       "      <td>NextSeq 550 sequencing; bulk RNA-seq of HCC139...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_title                                   experiment_title  \\\n",
       "0   HCC1395 BL  NextSeq 550 sequencing; bulk RNA-seq of HCC139...   \n",
       "1   HCC1395 BL  NextSeq 550 sequencing; bulk RNA-seq of HCC139...   \n",
       "2   HCC1395 BL  NextSeq 550 sequencing; bulk RNA-seq of HCC139...   \n",
       "3      HCC1395  NextSeq 550 sequencing; bulk RNA-seq of HCC139...   \n",
       "4      HCC1395  NextSeq 550 sequencing; bulk RNA-seq of HCC139...   \n",
       "5      HCC1395  NextSeq 550 sequencing; bulk RNA-seq of HCC139...   \n",
       "\n",
       "                         run_alias  library_source  read_count  \\\n",
       "0  HCC1395BL_1_S10_R1_001.fastq.gz  TRANSCRIPTOMIC    26549469   \n",
       "1  HCC1395BL_2_S11_R1_001.fastq.gz  TRANSCRIPTOMIC    28768620   \n",
       "2  HCC1395BL_3_S12_R1_001.fastq.gz  TRANSCRIPTOMIC    29246313   \n",
       "3    HCC1395_2_S14_R1_001.fastq.gz  TRANSCRIPTOMIC    28095730   \n",
       "4    HCC1395_3_S15_R1_001.fastq.gz  TRANSCRIPTOMIC    27883943   \n",
       "5    HCC1395_1_S13_R1_001.fastq.gz  TRANSCRIPTOMIC    30200567   \n",
       "\n",
       "  sample_description                                        description  \n",
       "0         HCC1395 BL  NextSeq 550 sequencing; bulk RNA-seq of HCC139...  \n",
       "1         HCC1395 BL  NextSeq 550 sequencing; bulk RNA-seq of HCC139...  \n",
       "2         HCC1395 BL  NextSeq 550 sequencing; bulk RNA-seq of HCC139...  \n",
       "3            HCC1395  NextSeq 550 sequencing; bulk RNA-seq of HCC139...  \n",
       "4            HCC1395  NextSeq 550 sequencing; bulk RNA-seq of HCC139...  \n",
       "5            HCC1395  NextSeq 550 sequencing; bulk RNA-seq of HCC139...  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bulk_samples[['sample_title', 'experiment_title', 'run_alias', 'library_source', 'read_count', 'sample_description', 'description']]#.loc[:, 'experiment_title'][1]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cbfe8694-6115-4e21-a72b-27e1145eae09",
   "metadata": {},
   "source": [
    "# prepare the convoluted samples for testing"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dd3e661e-2df3-4106-8ac2-82504ca16280",
   "metadata": {},
   "source": [
    "Each cell line has a single paired end library which has been sequenced three times.\n",
    "\n",
    "I will create three samples from the first replicate of HCC1395 with 10%, 50%, and 90% contamination from HCC1395 BL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "3d270cbc-9698-43ac-a2ef-87f7e4e4dd9a",
   "metadata": {},
   "outputs": [],
   "source": [
    "f_cancer = bulk_samples.loc[5].fastq_1.replace('/run/media/ian', '/home/jovyan')\n",
    "r_cancer = bulk_samples.loc[5].fastq_2.replace('/run/media/ian', '/home/jovyan')\n",
    "\n",
    "f_bcell = bulk_samples.loc[0].fastq_1.replace('/run/media/ian', '/home/jovyan')\n",
    "r_bcell = bulk_samples.loc[0].fastq_2.replace('/run/media/ian', '/home/jovyan')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "3bc0240c-78c7-4e7a-ad4c-19a8d7138aa7",
   "metadata": {},
   "outputs": [],
   "source": [
    "! mkdir -p ./output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "0c42ab99-df76-4ac3-8955-3a6f75fb83f4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 23.6 ms, sys: 169 ms, total: 193 ms\n",
      "Wall time: 14min 1s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "conv_portions = [.10, .50, .90]\n",
    "file_pairs = [[f_cancer, f_bcell, 1], [r_cancer, r_bcell, 2]]\n",
    "out_path = './output'\n",
    "for cp in conv_portions:\n",
    "    for fp in file_pairs:\n",
    "        file_name = str(cp * 100).replace('.0', 'Percent')\n",
    "        fA = fp[0]\n",
    "        fB = fp[1]\n",
    "        direction = str(fp[2])\n",
    "        output_fastq = join(out_path, 'conv_'+file_name+'_'+direction+'.fastq')\n",
    "        convolute_raw_data(fA, fB, output_fastq, cp)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cd82a70b-3fc2-421a-a852-36bd3453aa18",
   "metadata": {},
   "source": [
    "Check that new files have the correct number of reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "b91494d9-1138-40ad-8b01-2e8dc1e86aff",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 19.7 ms, sys: 49.4 ms, total: 69.1 ms\n",
      "Wall time: 1min 18s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "new_files = glob(out_path+'/*')\n",
    "p_conv = [10, 50, 90]\n",
    "for p in p_conv:\n",
    "    temp_list = [i for i in new_files if str(p) in i]\n",
    "    tf0 = temp_list[0]\n",
    "    tf1 = temp_list[1]\n",
    "    assert count_fastq_reads(tf0) == count_fastq_reads(tf1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3f87c4d1-f853-4da1-994c-de2e99f55311",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
